src/openssl.c

```diff
@@ -7474,6 +7462,65 @@ static int p12_interpose(lua_State *L) {
 } /* p12_interpose() */
 
 
+static int p12_parse(lua_State *L) {
+	/* parse a p12 binary string and return the parts */
+	lua_settop(L,2);
+
+	EVP_PKEY *pkey;
+	X509 *cert;
+	STACK_OF(X509) *ca = NULL;
+	PKCS12 *p12;
+
+	const char *blob, *passphrase;
+	size_t len;
+
+	blob = luaL_checklstring(L, 1, &len);
+	BIO *bio = BIO_new_mem_buf((void*)blob, len);
+	if (!bio) 
+		return auxL_error(L, auxL_EOPENSSL, "pkcs12.parse");
+	p12 = d2i_PKCS12_bio(bio, NULL);
+	BIO_free(bio);
+	if (!p12)
+		return auxL_error(L, auxL_EOPENSSL, "pkcs12.parse");
+
+	/* passphrase is the second parameter */
+	passphrase = luaL_optstring(L, 2, NULL);
```

Comment:
This (and the `prepsimple`s below) can longjump out, leaking `p12`. 
Do all the argument marshalling and userdata allocations upfront before any openssl allocations.